File: src/config/source/greengrass.rs
//! # Configuration source — GG_CONFIG (Phase 2)
//!
//! **One-liner purpose**: Load (and hot-reload) configuration from the Greengrass
//! deployment via IPC `GetConfiguration` + `SubscribeToConfigurationUpdate`.
//!
//! ## Overview
//! Reads the component's deployed configuration at a key path (default key
//! `ComponentConfig`) through the shared [`crate::ipc`] runtime, and watches it for
//! deployment-driven changes. Compiled only with the `greengrass` feature.
//!
//! ## Semantics & Architecture
//! - `load` issues `GetConfiguration` and returns the subtree as a JSON document.
//! - `watch` registers a `SubscribeToConfigurationUpdate`; because that operation
//!   delivers only the changed key path, the runtime re-fetches the value and
//!   forwards the fresh document on the channel.
//! - `Send + Sync`; async via `async_trait`. Errors map to
//!   [`crate::error::GgError::Ipc`].
//!
//! ## Status
//! Phase 2, **compile-only** — not yet validated against a live Greengrass core.
//!
//! ## Related Modules
//! - [`super`], [`crate::ipc`].

use async_trait::async_trait;
use serde_json::Value;
use tokio::sync::mpsc;

use super::ConfigSource;
use crate::error::Result;
use crate::ipc;

/// Greengrass-IPC-backed configuration source (`GetConfiguration`).
pub struct GreengrassConfigSource {
    /// Other component to read config from, or `None` for this component.
    component: Option<String>,
    /// Top-level configuration key to read (e.g. `ComponentConfig`).
    key: String,
}

impl GreengrassConfigSource {
    /// Create a source reading `key` (default `ComponentConfig`) from `component`
    /// (or this component when `None`).
    pub fn new(component: Option<String>, key: String) -> Self {
        Self { component, key }
    }

    /// The IPC key path: empty for the whole config, else the single configured key.
    fn key_path(&self) -> Vec<String> {
        if self.key.is_empty() {
            Vec::new()
        } else {
            vec![self.key.clone()]
        }
    }
}

#[async_trait]
impl ConfigSource for GreengrassConfigSource {
    async fn load(&self) -> Result<Value> {
        let rt = ipc::global();
        rt.connect().await?;
        rt.get_config(self.key_path(), self.component.clone()).await
    }

    fn source_name(&self) -> &str {
        "GG_CONFIG"
    }

    fn watch(&self) -> Option<mpsc::UnboundedReceiver<Value>> {
        let (tx, rx) = mpsc::unbounded_channel();
        let component = self.component.clone();
        let key_path = self.key_path();
        // Registration is async; spawn it. `build()` runs in an async context.
        tokio::spawn(async move {
            let rt = ipc::global();
            if let Err(e) = rt.connect().await {
                tracing::warn!(error = %e, "GG_CONFIG watch: connect failed");
                return;
            }
            if let Err(e) = rt.watch_config(component, key_path, tx).await {
                tracing::warn!(error = %e, "GG_CONFIG watch: subscribe failed");
            }
        });
        Some(rx)
    }
}

File: libs/ts/src/config/source/greengrass.ts
/**
 * Configuration source — GG_CONFIG.
 *
 * Loads (and hot-reloads) configuration from the Greengrass deployment via IPC
 * `GetConfiguration` + `SubscribeToConfigurationUpdate`. A port of the Rust
 * `greengrass.rs` source:
 *
 * - `load` issues `getConfiguration(keyPath, component)` and returns the subtree.
 * - The IPC key path is **empty** (whole config) when the configured `key` is empty,
 *   else the **single** configured key (it is NOT split on `.`) — matching Rust
 *   `key_path()`.
 * - `watch` registers `watchConfiguration`; because that operation delivers only the
 *   changed key path, the handler re-fetches the value and forwards the fresh doc.
 */
import { ConfigSource, ConfigWatch } from "./index";
import { IpcMessagingProvider } from "../../messaging/ipc-provider";

/** Greengrass-IPC-backed configuration source (`GetConfiguration`). */
export class GreengrassConfigSource implements ConfigSource {
  /**
   * @param ipc       the Greengrass IPC provider.
   * @param component other component to read config from, or `undefined` for this one.
   * @param key       top-level configuration key to read (e.g. `ComponentConfig`).
   */
  constructor(
    private readonly ipc: IpcMessagingProvider,
    private readonly component: string | undefined,
    private readonly key: string,
  ) {}

  /** The IPC key path: empty for the whole config, else the single configured key. */
  private keyPath(): string[] {
    return this.key === "" ? [] : [this.key];
  }

  async load(): Promise<unknown> {
    return this.ipc.getConfiguration(this.keyPath(), this.component);
  }

  sourceName(): string {
    return "GG_CONFIG";
  }

  async watch(onUpdate: (raw: unknown) => void): Promise<ConfigWatch | undefined> {
    const keyPath = this.keyPath();
    const sub = await this.ipc.watchConfiguration(keyPath, this.component, async () => {
      try {
        const value = await this.ipc.getConfiguration(keyPath, this.component);
        onUpdate(value);
      } catch (e) {
        console.warn(`GG_CONFIG watch: re-fetch failed: ${(e as Error).message}`);
      }
    });
    return {
      close: async () => {
        await sub.unsubscribe();
      },
    };
  }
}
